### PR TITLE
hot restart: block new starts if previous process is still initializing

### PR DESCRIPTION
**Pull request by mattklein123**
fixes https://github.com/lyft/envoy/issues/550

### CODE REVIEW

**Comment by mattklein123** [comment]
@lyft/network-team 

@kyessenov can you please test this PR and let me know if it fixes your issue.

**Comment by kyessenov** [comment]
I have trouble reproducing the race reliably. We run the fast restart on each pre-submit, so I can let you know if I this happens again after merge.

**Comment by mattklein123** [comment]
@kyessenov deploying a change like this at Lyft is kind of a pain. I would really like to know if this fixes your issue, and you are going to have to write code around it also to handle the error case. Isn't there some way you can test this locally?

**Comment by mattklein123** [comment]
@kyessenov has confirmed fix.

@lyft/network-team please review.

File: source/exe/hot_restart.cc [hunk 1]
```diff
@@ -53,6 +53,16 @@ SharedMemory& SharedMemory::initialize(Options& options) {
     RELEASE_ASSERT(shmem->version_ == VERSION);
   }
 
+  // Here we catch the case where a new Envoy starts up when the current Envoy has not yet fully
+  // initialized. The startup logic is quite compicated, and it's not worth trying to handle this
```
Comment (ccaraman) [inline source/exe/hot_restart.cc]: typo complicated

Comment (mattklein123) [inline source/exe/hot_restart.cc]: will fix in follow up.

**Pull request merged**